api/src/main/java/org/pmiops/workbench/mail/MailServiceImpl.java

```diff
@@ -114,25 +232,72 @@ public void sendBetaAccessCompleteEmail(final String contactEmail, final String
   }
 
   private Map<EmailSubstitutionField, String> betaAccessSubstitutionMap(final String username) {
-    final CloudStorageService cloudStorageService = cloudStorageServiceProvider.get();
-
+    final WorkbenchConfig workbenchConfig = workbenchConfigProvider.get();
     final String action =
         "login to the workbench via <a class=\"link\" href=\""
-            + workbenchConfigProvider.get().admin.loginUrl
+            + workbenchConfig.admin.loginUrl
             + "\">"
-            + workbenchConfigProvider.get().admin.loginUrl
+            + workbenchConfig.admin.loginUrl
             + "</a>";
 
     return new ImmutableMap.Builder<EmailSubstitutionField, String>()
         .put(EmailSubstitutionField.ACTION, action)
         .put(EmailSubstitutionField.BETA_ACCESS_REPORT, "approved for use")
         .put(
             EmailSubstitutionField.HEADER_IMG,
-            cloudStorageService.getImageUrl("all_of_us_logo.png"))
+            cloudStorageServiceProvider.get().getImageUrl("all_of_us_logo.png"))
         .put(EmailSubstitutionField.USERNAME, username)
         .build();
   }
 
+  private ImmutableMap<EmailSubstitutionField, String> freeTierDollarThresholdSubstitutionMap(
+      final DbUser user,
+      double currentUsage,
+      double remainingBalance,
+      final Optional<Instant> expirationTimeIfKnown) {
+
+    final String expirationDate =
+        expirationTimeIfKnown
+            .map(time -> formatDate(time.atZone(ZoneId.systemDefault()).toLocalDate()))
+            .orElse("[unknown]");
+
+    return new ImmutableMap.Builder<EmailSubstitutionField, String>()
+        .put(
+            EmailSubstitutionField.HEADER_IMG,
+            cloudStorageServiceProvider.get().getImageUrl("all_of_us_logo.png"))
+        .put(EmailSubstitutionField.FIRST_NAME, user.getGivenName())
+        .put(EmailSubstitutionField.LAST_NAME, user.getFamilyName())
+        .put(EmailSubstitutionField.USED_CREDITS, formatCurrency(currentUsage))
+        .put(EmailSubstitutionField.CREDIT_BALANCE, formatCurrency(remainingBalance))
+        .put(EmailSubstitutionField.EXPIRATION_DATE, expirationDate)
+        .build();
+  }
+
+  private ImmutableMap<EmailSubstitutionField, String> freeTierTimeThresholdSubstitutionMap(
+      DbUser user, long daysRemaining, double remainingDollarBalance, LocalDate expirationDate) {
+    return new ImmutableMap.Builder<EmailSubstitutionField, String>()
+        .put(
+            EmailSubstitutionField.HEADER_IMG,
+            cloudStorageServiceProvider.get().getImageUrl("all_of_us_logo.png"))
```

Comment:
nit: cloudStorageServiceProvider.get().getImageUrl("all_of_us_logo.png")) this is being repeated a lot of times, if its going to be the same i am wondering if it can be taken out, just so if we will ever change the image source we can do it at one place

Reply:
good idea, thanks